src/main/java/io/camunda/testing/assertions/ProcessInstanceAssertions.java

```diff
@@ -0,0 +1,413 @@
+   *
+   * @param elementIds The element ids
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions hasPassedElementInOrder(final String... elementIds) {
+    final List<String> foundElementRecords =
+        StreamFilter.processInstance(recordStreamSource.processInstanceRecords())
+            .withProcessInstanceKey(actual.getProcessInstanceKey()).withElementIdIn(elementIds)
+            .withIntent(ProcessInstanceIntent.ELEMENT_ACTIVATING).stream()
+            .map(Record::getValue)
+            .map(ProcessInstanceRecordValue::getElementId)
+            .collect(Collectors.toList());
+
+    assertThat(foundElementRecords)
+        .describedAs("Ordered elements")
+        .isEqualTo(Arrays.asList(elementIds));
+
+    return this;
+  }
+
+  /**
+   * Verifies the expectation that the process instance is currently waiting at one or more
+   * specified elements.
+   *
+   * @param elementIds The ids of the elements
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions isWaitingAtElement(final String... elementIds) {
+    final List<String> notActivatedElements =
+        Arrays.stream(elementIds)
+            .filter(((Predicate<String>) this::isWaitingAtElement).negate())
+            .collect(Collectors.toList());
+
+    assertThat(notActivatedElements.isEmpty())
+        .withFailMessage(
+            "Process with key %s is not waiting at element(s) with id(s) %s",
+            actual.getProcessInstanceKey(), String.join(", ", notActivatedElements))
+        .isTrue();
+
+    return this;
+  }
+
+  /**
+   * Verifies the expectation that the process instance is currently not waiting at one or more
+   * specified elements.
+   *
+   * @param elementIds The ids of the elements
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions isNotWaitingAtElement(final String... elementIds) {
+    final List<String> activatedElements =
+        Arrays.stream(elementIds).filter(this::isWaitingAtElement).collect(Collectors.toList());
+
+    assertThat(activatedElements.isEmpty())
+        .withFailMessage(
+            "Process with key %s is waiting at element(s) with id(s) %s",
+            actual.getProcessInstanceKey(), String.join(", ", activatedElements))
+        .isTrue();
+
+    return this;
+  }
+
+  /**
+   * Checks if the process instance is currently waiting at an element with a specific element id.
+   *
+   * @param elementId The id of the element
+   * @return boolean indicating whether the process instance is waiting at the element
+   */
+  // TODO if waiting at any element with this id returns true. Maybe add a counter so we can check
+  // we are waiting at multiple elements with the same id
+  private boolean isWaitingAtElement(final String elementId) {
+    final Map<Long, List<Record<ProcessInstanceRecordValue>>> recordsMap =
```

Comment:
:thought_balloon: This method is difficult to read. I think it is correct, though.

I wonder whether the first map can be reused for other methods in here as well.

Reply:
:thought_balloon: I think we could also discuss this with @npepinpe and @saig0 to make sure we have the same semantics of "waiting" in mind.

Reply:
:x: Come to think of it, we generally have to check that the command was not rejected

`record -> record.getRejectionType() == RejectionType.NULL_VAL`

Reply:
> make sure we have the same semantics of "waiting" in mind.

I think the term "waiting on an element" is common. So, an element is entered but not yet left. 

It is not precise. In Zeebe, we would say that an element (instance) is activated. Or, it is completed. 
In the docs, we use both "entered" and "activated".

But I guess for non-technical users, the term "waiting" sounds more familiar.

Reply:
My question was more related to what do we define as wait state? 

Right now we use "the last record for an element was intent = ELEMENT_ACTIVATED" as filter criterion for wait state.
What about situations where it is "waiting" because an incident was raised? Is that waiting, or is that something else?

Reply:
> What about situations where it is "waiting" because an incident was raised? Is that waiting, or is that something else?

Good point. 

In Operate, the element should be marked as active. AFAIK, it is active from `activating` to `completed`. 

In general, it would say that an element reached the wait state when it is `activated`. If an incident occurs before then it is not waiting - because you can't trigger the element (e.g. complete the job, send a message, etc.). Same if an incident occurs after `completing` then the element is already on leave.